node/src/components/fetcher.rs

```diff
@@ -467,14 +476,26 @@ impl ItemFetcher<BlockWithMetadata> for Fetcher<BlockWithMetadata> {
         peer: NodeId,
         responder: FetchResponder<BlockWithMetadata>,
     ) -> Effects<Event<BlockWithMetadata>> {
-        effect_builder
-            .get_block_and_sufficient_finality_signatures_by_height_from_storage(id)
-            .event(move |result| Event::GetFromStorageResult {
-                id,
-                peer,
-                maybe_item: Box::new(result),
-                responder,
-            })
+        let finality_threshold_fraction = self.finality_threshold_fraction;
+        async move {
+            let block_with_metadata = effect_builder
+                .get_block_with_metadata_from_storage_by_height(id, false)
+                .await?;
+            has_enough_block_signatures(
+                effect_builder,
+                block_with_metadata.block.header(),
+                &block_with_metadata.finality_signatures,
+                finality_threshold_fraction,
+            )
+            .await
+            .then_some(block_with_metadata)
+        }
+        .event(move |result| Event::GetFromStorageResult {
+            id,
+            peer,
+            maybe_item: Box::new(result),
```

Comment:
Might be better to make this an `Option<Box<_>>` instead of `Box<Option<_>>`, to avoid allocation in the `None` case.

Reply:
Agreed, but too big of a change for this PR I think.